File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int[] arr;
    int targetNumber, answer=0;
    
    public void findTargetNum(int num, int cnt){
        if(num==targetNumber&&cnt==arr.length){
            answer++;
            return;
        }
        if(cnt>=arr.length) return;
        
        findTargetNum(num-arr[cnt],cnt+1);
        findTargetNum(num+arr[cnt],cnt+1);
    }    
    public int solution(int[] numbers, int target) {
        int n=numbers.length;
        arr=new int[n];
        targetNumber=target;
        
        for(int i=0;i<n;i++){
            arr[i]=numbers[i];
        }
        
        findTargetNum(-1*arr[0],1);
        findTargetNum(arr[0],1);
        return answer;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# def solution(numbers, target):
#     def dfs(cnt, current_sum):
#         if cnt == len(numbers):
#             if current_sum == target:
#                 return 1
#             else:
#                 return 0
#         return dfs(cnt + 1, current_sum + numbers[cnt]) + dfs(cnt + 1, current_sum - numbers[cnt])
#     return dfs(0, 0)


def solution(numbers, target):
    def dfs(idx, total):
        if idx==len(numbers):
            if total==target:
                return 1
            else:
                return 0
        return dfs(idx+1, total+numbers[idx])+dfs(idx+1, total-numbers[idx])
    return dfs(0,0)